### PR TITLE
fabtest/rdm_stress: Add stress test for RDM endpoints

### PR DESCRIPTION
**Pull request by shefty**
Stress multi-threaded and multi-process access to an RDM endpoint
with abrupt termination conditions.  This will help verify or
expose race conditions in error handling cases.

This patch defines the general framework of operation and
implements most of the client code.  The server side is still
needed.

The general operation is for the client to parse an input file
describing a set of RPCs to perform.  Each RPC involves sending
a short message to the server.  The server will respond by either
sending a message, a tagged message, or issuing an RMA read or
write operation back to the client.  RMA operations are followed
by a response message.

The client side has a parent process and several child processes.
Child processes are forked from the parent, with each child
executing a set of RPC calls to a single server.  The parent
spawns new children as children processes exit.  A child process
may not wait for an RPC response and could exit after sending
a request.

The intent for the server is that the main thread will receive
RPC messages and spawn a thread to handle that request.  The
new thread will generate the RPC response and exit.

The objective is to test a multi-threaded server under load
from multiple clients which are possibly misbehaving.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

### CODE REVIEW

File: fabtests/functional/rdm_stress.c [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+	int i, ret;
+
+	ret = init_rpc();
+	if (ret)
+		return ret;
+
+	for (i = 0; i < opts.iterations; i++) {
+		for (myid = 0; myid < opts.num_connections; myid++) {
+			if (clients[myid].pid) {
+				waitpid(clients[myid].pid, NULL, 0);
```
Comment (j-xiong) [inline fabtests/functional/rdm_stress.c:406]: Would it be better to use wait() or waitpid(-1, ..) to catch any child process instead wait for them in turn?

File: fabtests/functional/rdm_stress.c [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+#include <getopt.h>
+#include <sys/types.h>
+#include <wait.h>
+#include <unistd.h>
+
+#include <rdma/fi_cm.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_tagged.h>
+#include <shared.h>
```
[the remark this answers is not in the document]
Comment (shefty) [reply fabtests/functional/rdm_stress.c:42]: For fabtests, these are installed headers.

[the remark this answers is not in the document]
Comment (shefty) [reply fabtests/functional/rdm_stress.c:42]: Oh - I think you're only referring to shared.h.  I'll update

[the remark this answers is not in the document]
Comment (j-xiong) [reply fabtests/functional/rdm_stress.c:42]: Yes, only the single line.

**Comment by shefty** [comment]
Added 2 patches to the av implementation.  The first adds a print if a duplicate address is inserted.  The second restructures the rxm av remove code to optimize the locking and avoid issues if a duplicate address is inserted (depending on the app's behavior).  Those patches are both candidates to pull into v1.15.

The rdm_stress program is now working, with the exception that the input file parsing is commented out.  Instead I have a hard-coded 'input' array that I am using to test and debug the framework.  That array will be removed once the parsing code is re-enabled.  I also changed the input format (simplified it), so the parsing code needs to be updated.

The good news is that the stress program appears to be doing the job of exposing issues.  The bad news is that I don't know if the issues are in the program or libfabric, but I believe the latter.

**Pull request merged**